datahub/user/company_list/tests/test_views.py

```diff
@@ -85,3 +101,187 @@ def test_lists_are_sorted_by_name(self):
+        response = self.api_client.put(url)
+
+        assert response.status_code == status.HTTP_204_NO_CONTENT
+        assert response.content == b''
+
+        assert CompanyListItem.objects.filter(
+            list__in=(other_list, company_list),
+            company=company,
+        ).count() == 2
+
+    def test_with_existing_item(self):
+        """
+        Test that no error is returned if the specified company is already on the
+        authenticated user's list.
+        """
+        creation_date = datetime(2018, 1, 2, tzinfo=utc)
+        modified_date = datetime(2018, 1, 2, tzinfo=utc)
+        company = CompanyFactory()
+
+        company_list = CompanyListFactory(adviser=self.user)
+
+        with freeze_time(creation_date):
+            CompanyListItemFactory(list=company_list, company=company)
+
+        url = _get_list_item_url(company_list.pk, company.pk)
+
+        with freeze_time(modified_date):
+            response = self.api_client.put(url)
+
+        assert response.status_code == status.HTTP_204_NO_CONTENT
+        assert response.content == b''
+
+        company_list_item = _get_queryset_for_list_adviser_and_company(
+            company_list,
+            company,
+        ).first()
+
+        assert company_list_item.created_on == creation_date
+        assert company_list_item.modified_on == modified_date
+
+    def test_with_archived_company(self):
+        """Test that an archived company can't be added to the authenticated user's list."""
+        company_list = CompanyListFactory(adviser=self.user)
+        company = ArchivedCompanyFactory()
+
+        url = _get_list_item_url(company_list.pk, company.pk)
+        response = self.api_client.put(url)
+
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+        assert response.json() == {
+            api_settings.NON_FIELD_ERRORS_KEY: CANT_ADD_ARCHIVED_COMPANY_MESSAGE,
+        }
+
+    def test_with_non_existent_company(self):
+        """Test that a 404 is returned if the specified company ID is invalid."""
+        company_list = CompanyListFactory(adviser=self.user)
+
+        url = _get_list_item_url(company_list.pk, uuid4())
+        response = self.api_client.put(url)
+
+        assert response.status_code == status.HTTP_404_NOT_FOUND
+
+    def test_with_non_existent_list(self):
+        """Test that you cannot add an item to a list that does not exist."""
+        # Existing company lists for other users should not matter
+        CompanyListFactory.create_batch(5)
+        company = CompanyFactory()
+
+        url = _get_list_item_url(uuid4(), company.pk)
+        response = self.api_client.put(url)
+
+        assert response.status_code == status.HTTP_404_NOT_FOUND
+        assert not CompanyListItem.objects.filter(
+            list__adviser=self.user,
+            company=company,
+        ).exists()
+
+
+def _get_queryset_for_list_adviser_and_company(company_list, company):
```

Comment:
I think we can drop 'adviser from the function name now:

```suggestion
def _get_queryset_for_list_and_company(company_list, company):
```